util/src/main/java/io/kubernetes/client/Exec.java

```diff
@@ -170,20 +177,65 @@ public Process exec(
       throws ApiException, IOException {
     String path = makePath(namespace, name, command, container, stdin, tty);
 
-    WebSocketStreamHandler handler = new WebSocketStreamHandler();
-    ExecProcess exec = new ExecProcess(handler);
+    ExecProcess exec = new ExecProcess();
+    WebSocketStreamHandler handler = exec.getHandler();
     WebSockets.stream(path, "GET", apiClient, handler);
 
     return exec;
   }
 
+  static int parseExitCode(InputStream inputStream) {
+    int exitCode = 0;
+    try {
+      int available = inputStream.available();
+      if (available > 0) {
+        byte[] b = new byte[available];
+        inputStream.read(b);
+        String result = new String(b, "UTF-8");
+        int idx = result.lastIndexOf(':');
+        if (idx > 0) {
+          try {
+            exitCode = Integer.parseInt(result.substring(idx + 1).trim());
+          } catch (NumberFormatException nfe) {
+            log.error("Error parsing exit code from status channel response", nfe);
+          }
+        }
+      }
+    } catch (IOException io) {
+      log.error("Error parsing exit code from status channel response", io);
+    }
+
+    return exitCode;
+  }
+
   private static class ExecProcess extends Process {
-    WebSocketStreamHandler streamHandler;
-    private int statusCode;
+    private final WebSocketStreamHandler streamHandler;
+    private volatile int statusCode;
+    private final Map<Integer, InputStream> input = new HashMap<>();
 
-    public ExecProcess(WebSocketStreamHandler handler) throws IOException {
-      this.streamHandler = handler;
+    public ExecProcess() throws IOException {
       this.statusCode = -1;
+      this.streamHandler =
+          new WebSocketStreamHandler() {
+            @Override
+            public void close() {
+              if (statusCode == -1) {

```

Comment:
I feel like we may want to loop here in case the network stream is blocked up or otherwise unavailable.  Obviously probably need to time out eventually...

Reply:
I checked the Web Socket specification here: https://datatracker.ietf.org/doc/rfc6455/?include_text=1.  The server has sent a Close message.  This is why there is the callback to WebSocketListener.onClose(), which is then calling this close() method.

Therefore, we know that there will be no more calls to onMessage().